File: ros2_system/weight_detection_module/weight_detection_module/weight_detector.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import JointState
from std_msgs.msg import Float32
import numpy as np
from collections import deque
import matplotlib.pyplot as plt

class KalmanFilter:
    def __init__(self, process_variance=0.0015, measurement_variance=0.04):
        self.process_variance = process_variance
        self.measurement_variance = measurement_variance
        self.estimate = 0.0
        self.error_covariance = 1.0
        self.initialized = False
    
    def update(self, measurement):
        if not self.initialized:
            self.estimate = measurement
            self.initialized = True
            return self.estimate
        
        self.error_covariance += self.process_variance
        kalman_gain = self.error_covariance / (self.error_covariance + self.measurement_variance)
        self.estimate += kalman_gain * (measurement - self.estimate)
        self.error_covariance *= (1 - kalman_gain)
        
        return self.estimate

class UR5eKinematics:
    def __init__(self):
        self.d1 = 0.1625
        self.a2 = -0.425
        self.a3 = -0.3922
        self.d4 = 0.1333
        self.d5 = 0.0997
        self.d6 = 0.0996
        self.g = 9.81
    
    def compute_dh_transform(self, theta, d, a, alpha):
        ct = np.cos(theta)
        st = np.sin(theta)
        ca = np.cos(alpha)
        sa = np.sin(alpha)
        
        return np.array([
            [ct, -st*ca, st*sa, a*ct],
            [st, ct*ca, -ct*sa, a*st],
            [0, sa, ca, d],
            [0, 0, 0, 1]
        ])
    
    def forward_kinematics(self, joint_angles):
        q = joint_angles
        
        T01 = self.compute_dh_transform(q[0], self.d1, 0, np.pi/2)
        T12 = self.compute_dh_transform(q[1], 0, self.a2, 0)
        T23 = self.compute_dh_transform(q[2], 0, self.a3, 0)
        T34 = self.compute_dh_transform(q[3], self.d4, 0, np.pi/2)
        T45 = self.compute_dh_transform(q[4], self.d5, 0, -np.pi/2)
        T56 = self.compute_dh_transform(q[5], self.d6, 0, 0)
        
        T02 = T01 @ T12
        T03 = T02 @ T23
        T04 = T03 @ T34
        T05 = T04 @ T45
        T06 = T05 @ T56
        
        joint_positions = {
            'shoulder': T01[:3, 3],
            'elbow': T02[:3, 3],
            'wrist1': T03[:3, 3],
            'end_effector': T06[:3, 3]
        }
        
        return joint_positions, T01
    
    def compute_moment_arms(self, joint_angles):
        positions, T01 = self.forward_kinematics(joint_angles)
        ee_pos = positions['end_effector']
        
        joint2_pos = positions['shoulder']
        joint3_pos = positions['elbow']
        joint4_pos = positions['wrist1']
        
        joint2_axis = np.array([0, 0, 1])
        joint3_axis = T01[:3, :3] @ joint2_axis
        joint4_axis = joint3_axis
        
        vec_to_ee_j2 = ee_pos - joint2_pos
        vec_to_ee_j3 = ee_pos - joint3_pos
        vec_to_ee_j4 = ee_pos - joint4_pos
        
        moment_j2 = np.linalg.norm(np.cross(joint2_axis, vec_to_ee_j2))
        moment_j3 = np.linalg.norm(np.cross(joint3_axis, vec_to_ee_j3))
        moment_j4 = np.linalg.norm(np.cross(joint4_axis, vec_to_ee_j4))
        
        return np.array([moment_j2, moment_j3, moment_j4])
    
    def estimate_mass(self, torque_deltas, moment_arms):
        valid_indices = np.where(np.abs(moment_arms) > 0.01)[0]
        
        if len(valid_indices) == 0:
            return 0.0
        
        mass_estimates = []
        for idx in valid_indices:
            mass = torque_deltas[idx] / (self.g * moment_arms[idx])
            mass_estimates.append(mass)
        
        return np.median(mass_estimates)

class WeightDetector(Node):
    def __init__(self):
        super().__init__('weight_detector')
        
        self.subscription = self.create_subscription(
            JointState,
            '/joint_states',
            self.joint_state_callback,
            10
        )
        
        self.mass_publisher = self.create_publisher(Float32, '/estimated_mass', 10)
        
        self.num_joints = 6
        self.history_length = 100
        self.torque_history = [deque(maxlen=self.history_length) for _ in range(self.num_joints)]
        self.filtered_torque_history = [deque(maxlen=self.history_length) for _ in range(self.num_joints)]
        
        self.kalman_filters = [KalmanFilter(process_variance=0.0015, measurement_variance=0.04) 
                               for _ in range(self.num_joints)]
        
        self.kinematics = UR5eKinematics()
        
        self.calibration_factor = 5.0
        self.active_joints = [0, 1]
        
        self.baseline_torques = None
        self.baseline_samples = []
        self.baseline_sample_size = 10
        self.calibrating_baseline = True
        
        self.current_joint_angles = None
        self.estimated_mass_grams = 0.0
        self.mass_history = deque(maxlen=self.history_length)
        
        self.declare_parameter('calibration_factor', self.calibration_factor)
        self.declare_parameter('active_joints', self.active_joints)
        
        self.param_timer = self.create_timer(0.5, self.update_parameters)
        
        plt.ion()
        self.fig = plt.figure(figsize=(14, 10))
        gs = self.fig.add_gridspec(3, 3, hspace=0.3)
        
        self.axes = []
        for i in range(6):
            row = i // 3
            col = i % 3
            ax = self.fig.add_subplot(gs[row, col])
            self.axes.append(ax)
        
        self.mass_ax = self.fig.add_subplot(gs[2, :])
        
        self.raw_lines = []
        self.filtered_lines = []
        
        for i in range(self.num_joints):
            raw_line, = self.axes[i].plot([], [], 'b-', alpha=0.3, label='Raw')
            filtered_line, = self.axes[i].plot([], [], 'r-', linewidth=2, label='Filtered')
            self.raw_lines.append(raw_line)
            self.filtered_lines.append(filtered_line)
            self.axes[i].set_xlabel('Sample')
            self.axes[i].set_ylabel('Torque (Nm)')
            self.axes[i].set_title(f'Joint {i+1}')
            self.axes[i].grid(True)
            self.axes[i].legend()
        
        self.mass_line, = self.mass_ax.plot([], [], 'g-', linewidth=2)
        self.mass_ax.set_xlabel('Sample')
        self.mass_ax.set_ylabel('Mass (g)')
        self.mass_ax.set_title('Estimated Mass')
        self.mass_ax.grid(True)
        
        self.get_logger().info('Weight detection module initialized - calibrating baseline...')
        self.get_logger().info('Using proper 3D moment arm calculations')
        self.get_logger().info('Note: UR5e joints reordered from [6,1,2,3,4,5] to [1,2,3,4,5,6] for display')
    
    def update_parameters(self):
        self.calibration_factor = self.get_parameter('calibration_factor').value
        self.active_joints = self.get_parameter('active_joints').value
    
    def joint_state_callback(self, msg):
        joint_torques = msg.effort[:self.num_joints] if len(msg.effort) >= self.num_joints else msg.effort
        joint_positions = msg.position[:self.num_joints] if len(msg.position) >= self.num_joints else msg.position
        
        if len(joint_torques) < self.num_joints or len(joint_positions) < self.num_joints:
            return
        
        # UR5e returns joints in order: [6, 1, 2, 3, 4, 5] (indices [5, 0, 1, 2, 3, 4])
        # Reorder to standard [1, 2, 3, 4, 5, 6] for display and physics
        reorder_indices = [5, 0, 1, 2, 3, 4]
        joint_torques_reordered = [joint_torques[i] for i in reorder_indices]
        joint_positions_reordered = [joint_positions[i] for i in reorder_indices]
        
        self.current_joint_angles = np.array(joint_positions_reordered)
        
        filtered_torques = []
        for i in range(self.num_joints):
            filtered_torque = self.kalman_filters[i].update(joint_torques_reordered[i])
            self.torque_history[i].append(joint_torques_reordered[i])
            self.filtered_torque_history[i].append(filtered_torque)
            filtered_torques.append(filtered_torque)
        
        # Use joints 2, 3, 4 (indices 1, 2, 3 after reordering)
        if self.calibrating_baseline:
            self.baseline_samples.append(filtered_torques[1:4])
            if len(self.baseline_samples) >= self.baseline_sample_size:
                self.baseline_torques = np.mean(self.baseline_samples, axis=0)
                self.calibrating_baseline = False
                self.get_logger().info("Baseline calibration complete")
                self.get_logger().info(f"Baseline torques for Joints 2,3,4: {self.baseline_torques}")
        else:
            self.estimate_mass_physics(filtered_torques[1:4])
        
        self.update_plots()
    
    def estimate_mass_physics(self, current_torques):
        if self.baseline_torques is None or self.current_joint_angles is None:
            return
        
        torque_deltas = np.array(current_torques) - self.baseline_torques
        
        moment_arms = self.kinematics.compute_moment_arms(self.current_joint_angles)
        
        selected_torque_deltas = torque_deltas[self.active_joints]
        selected_moment_arms = moment_arms[self.active_joints]
        
        raw_mass = self.kinematics.estimate_mass(selected_torque_deltas, selected_moment_arms)
        
        self.estimated_mass_grams = raw_mass * self.calibration_factor * 1000
        
        self.estimated_mass_grams = round(self.estimated_mass_grams / 5) * 5
        
        self.mass_history.append(self.estimated_mass_grams)
        
        mass_msg = Float32()
        mass_msg.data = self.estimated_mass_grams / 1000.0
        self.mass_publisher.publish(mass_msg)
    
    def update_plots(self):
        for i in range(self.num_joints):
            if len(self.torque_history[i]) > 0:
                self.raw_lines[i].set_data(range(len(self.torque_history[i])), 
                                          list(self.torque_history[i]))
                self.filtered_lines[i].set_data(range(len(self.filtered_torque_history[i])), 
                                               list(self.filtered_torque_history[i]))
                
                self.axes[i].relim()
                self.axes[i].autoscale_view()
        
        if len(self.mass_history) > 0:
            self.mass_line.set_data(range(len(self.mass_history)), 
                                   list(self.mass_history))
            self.mass_ax.relim()
            self.mass_ax.autoscale_view()
            
            self.mass_ax.set_title(f'Estimated Mass (Calib={self.calibration_factor:.2f}): {self.estimated_mass_grams:.2f}g')
        
        self.fig.canvas.draw()
        self.fig.canvas.flush_events()

def main(args=None):
    rclpy.init(args=args)
    weight_detector = WeightDetector()
    
    try:
        rclpy.spin(weight_detector)
    except KeyboardInterrupt:
        pass
    finally:
        weight_detector.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: ros2_system/src/weight_detection_module/src/weight_detector.cpp
// weight_detector.cpp - C++ implementation of weight detection node
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/joint_state.hpp>
#include <std_msgs/msg/int32.hpp>
#include <std_msgs/msg/bool.hpp>
#include <sort_interfaces/srv/calibrate_baseline.hpp>
#include <Eigen/Dense>
#include <deque>
#include <vector>
#include <cmath>
#include <algorithm>
#include <numeric>
#include <chrono>
#include <map>

class KalmanFilter
{
public:
    KalmanFilter(double process_variance = 0.001, double measurement_variance = 0.15)
        : process_variance_(process_variance),
          measurement_variance_(measurement_variance),
          estimate_(0.0),
          error_covariance_(1.0),
          initialized_(false) {}

    double update(double measurement)
    {
        if (!initialized_)
        {
            estimate_ = measurement;
            initialized_ = true;
            return estimate_;
        }

        error_covariance_ += process_variance_;
        double kalman_gain = error_covariance_ / (error_covariance_ + measurement_variance_);
        estimate_ += kalman_gain * (measurement - estimate_);
        error_covariance_ *= (1 - kalman_gain);

        return estimate_;
    }

    void reset()
    {
        estimate_ = 0.0;
        error_covariance_ = 1.0;
        initialized_ = false;
    }

private:
    double process_variance_;
    double measurement_variance_;
    double estimate_;
    double error_covariance_;
    bool initialized_;
};

class UR5eKinematics
{
public:
    UR5eKinematics() : g_(9.81)
    {
        // UR5e DH parameters
        d1_ = 0.1625;
        a2_ = -0.425;
        a3_ = -0.3922;
        d4_ = 0.1333;
        d5_ = 0.0997;
        d6_ = 0.0996;
    }

    Eigen::Matrix4d computeDHTransform(double theta, double d, double a, double alpha)
    {
        double ct = std::cos(theta);
        double st = std::sin(theta);
        double ca = std::cos(alpha);
        double sa = std::sin(alpha);

        Eigen::Matrix4d T;
        T << ct, -st * ca, st * sa, a * ct,
            st, ct * ca, -ct * sa, a * st,
            0, sa, ca, d,
            0, 0, 0, 1;
        return T;
    }

    std::pair<std::map<std::string, Eigen::Vector3d>, Eigen::Matrix4d>
    forwardKinematics(const std::vector<double> &q)
    {
        Eigen::Matrix4d T01 = computeDHTransform(q[0], d1_, 0, M_PI / 2);
        Eigen::Matrix4d T12 = computeDHTransform(q[1], 0, a2_, 0);
        Eigen::Matrix4d T23 = computeDHTransform(q[2], 0, a3_, 0);
        Eigen::Matrix4d T34 = computeDHTransform(q[3], d4_, 0, M_PI / 2);
        Eigen::Matrix4d T45 = computeDHTransform(q[4], d5_, 0, -M_PI / 2);
        Eigen::Matrix4d T56 = computeDHTransform(q[5], d6_, 0, 0);

        Eigen::Matrix4d T02 = T01 * T12;
        Eigen::Matrix4d T03 = T02 * T23;
        Eigen::Matrix4d T04 = T03 * T34;
        Eigen::Matrix4d T05 = T04 * T45;
        Eigen::Matrix4d T06 = T05 * T56;

        std::map<std::string, Eigen::Vector3d> positions;
        positions["shoulder"] = T01.block<3, 1>(0, 3);
        positions["elbow"] = T02.block<3, 1>(0, 3);
        positions["wrist1"] = T03.block<3, 1>(0, 3);
        positions["end_effector"] = T06.block<3, 1>(0, 3);

        return {positions, T01};
    }

    Eigen::Vector3d computeMomentArms(const std::vector<double> &joint_angles)
    {
        auto [positions, T01] = forwardKinematics(joint_angles);
        Eigen::Vector3d ee_pos = positions["end_effector"];

        Eigen::Vector3d joint2_pos = positions["shoulder"];
        Eigen::Vector3d joint3_pos = positions["elbow"];
        Eigen::Vector3d joint4_pos = positions["wrist1"];

        Eigen::Vector3d joint2_axis(0, 0, 1);
        Eigen::Vector3d joint3_axis = T01.block<3, 3>(0, 0) * joint2_axis;
        Eigen::Vector3d joint4_axis = joint3_axis;

        Eigen::Vector3d vec_to_ee_j2 = ee_pos - joint2_pos;
        Eigen::Vector3d vec_to_ee_j3 = ee_pos - joint3_pos;
        Eigen::Vector3d vec_to_ee_j4 = ee_pos - joint4_pos;

        double moment_j2 = joint2_axis.cross(vec_to_ee_j2).norm();
        double moment_j3 = joint3_axis.cross(vec_to_ee_j3).norm();
        double moment_j4 = joint4_axis.cross(vec_to_ee_j4).norm();

        return Eigen::Vector3d(moment_j2, moment_j3, moment_j4);
    }

    double estimateMass(const Eigen::VectorXd &torque_deltas,
                        const Eigen::VectorXd &moment_arms)
    {
        std::vector<double> mass_estimates;

        for (int i = 0; i < torque_deltas.size(); ++i)
        {
            if (std::abs(moment_arms(i)) > 0.01)
            {
                double mass = torque_deltas(i) / (g_ * moment_arms(i));
                mass_estimates.push_back(mass);
            }
        }

        if (mass_estimates.empty())
            return 0.0;

        // Return median
        std::nth_element(mass_estimates.begin(),
                         mass_estimates.begin() + mass_estimates.size() / 2,
                         mass_estimates.end());
        return mass_estimates[mass_estimates.size() / 2];
    }

    double getGravity() const { return g_; }

private:
    double g_;
    double d1_, a2_, a3_, d4_, d5_, d6_;
};

class WeightDetector : public rclcpp::Node
{
public:
    WeightDetector() : Node("weight_detector_cpp"),
                       calibrating_baseline_(true),
                       calibration_factor_(5.0),
                       estimated_mass_grams_(0.0),
                       messages_received_(0),
                       callback_counter_(0),
                       process_decimation_(5)
    {

        // Parameters
        this->declare_parameter("active_joints", std::vector<int64_t>{1, 2}); // Joints 1 -> 6 (0-indexed), {1,2} = joints 2 and 3
        this->declare_parameter("process_decimation", 10);                    // Process every Nth message (500Hz/10 = 50Hz)
        this->declare_parameter("baseline_sample_size", 220);                 // At decimation=10, ~5.5 seconds at 500Hz

        active_joints_ = this->get_parameter("active_joints").as_integer_array();
        baseline_sample_size_ = this->get_parameter("baseline_sample_size").as_int();
        process_decimation_ = this->get_parameter("process_decimation").as_int();

        // Snapping mode toggle: when true, uses exponential calibration and
        // snaps to weight set (0, 50, 100, 200, 500)
        // When false, uses polynomial calibration and rounds to nearest 5g
        this->declare_parameter("useSnapping", true);
        use_snapping_ = this->get_parameter("useSnapping").as_bool();

        // Exponential calibration parameters
        exp_amplitude_light_ = 15;
        exp_amplitude_heavy_ = 8.45;
        decay_light_ = 5.75;
        decay_heavy_ = 2.95;

        // Thresholds
        mass_threshold_ = 0.3; // kg, used to switch between calibration gains (in both exponential & polynomial)
        mass_threshold_ /= calibration_factor_;
        min_threshold_ = 0.0175; // kg, values below this are ignored
        min_threshold_ /= calibration_factor_;

        // Polynomial calibration parameters
        this->declare_parameter("poly_coeff_a", 0.0);     // cubic term (unused)
        this->declare_parameter("poly_coeff_b", -0.1463); // quadratic term
        this->declare_parameter("poly_coeff_c", 19.55);   // linear term
        this->declare_parameter("poly_coeff_d", -152.7);  // constant term

        poly_coeff_a_ = this->get_parameter("poly_coeff_a").as_double();
        poly_coeff_b_ = this->get_parameter("poly_coeff_b").as_double();
        poly_coeff_c_ = this->get_parameter("poly_coeff_c").as_double();
        poly_coeff_d_ = this->get_parameter("poly_coeff_d").as_double();

        // Weight set for snapping
        weight_set_ = {50, 100, 200, 500};

        // Kalman filters (6 joints)
        for (int i = 0; i < 6; ++i)
        {
            double process_variance = 0.0005;
            double measurement_variance = 15;
            kalman_filters_.emplace_back(process_variance, measurement_variance);
        }

        // QoS for sensor data
        auto qos = rclcpp::SensorDataQoS();
        qos.keep_last(1);

        // Subscription to UR5e Joint State Data
        joint_state_sub_ = this->create_subscription<sensor_msgs::msg::JointState>(
            "/joint_states",
            qos,
            std::bind(&WeightDetector::jointStateCallback, this, std::placeholders::_1));

        // Publisher of estimated mass
        mass_pub_ = this->create_publisher<std_msgs::msg::Int32>("/estimated_mass", 10);
        calibration_status_pub_ = this->create_publisher<std_msgs::msg::Bool>(
            "/weight_detection/calibration_status", 10);

        // Calibration service call
        calibration_service_ = this->create_service<sort_interfaces::srv::CalibrateBaseline>(
            "/weight_detection/calibrate_baseline",
            std::bind(&WeightDetector::calibrateBaselineCallback, this,
                      std::placeholders::_1, std::placeholders::_2));

        // Diagnostic timer
        diagnostic_timer_ = this->create_wall_timer(
            std::chrono::seconds(5),
            std::bind(&WeightDetector::reportDiagnostics, this));

        // Parameter update timer
        param_timer_ = this->create_wall_timer(
            std::chrono::milliseconds(500),
            std::bind(&WeightDetector::updateParameters, this));

        RCLCPP_INFO(this->get_logger(),
                    "Weight detector C++ node initialized - calibrating baseline...");
        RCLCPP_INFO(this->get_logger(),
                    "Active joints for mass estimation: [%ld, %ld]",
                    active_joints_.size() > 0 ? active_joints_[0] + 1 : 0,
                    active_joints_.size() > 1 ? active_joints_[1] + 1 : 0);
        RCLCPP_INFO(this->get_logger(),
                    "Snapping mode: %s", use_snapping_ ? "ENABLED" : "DISABLED");
    }

private:
    void updateParameters()
    {
        active_joints_ = this->get_parameter("active_joints").as_integer_array();
    }

    void calibrateBaselineCallback(
        const std::shared_ptr<sort_interfaces::srv::CalibrateBaseline::Request> /*request*/,
        std::shared_ptr<sort_interfaces::srv::CalibrateBaseline::Response> response)
    {

        RCLCPP_INFO(this->get_logger(),
                    "Baseline recalibration requested - resetting baseline and Kalman filters...");

        baseline_torques_.clear();
        baseline_samples_.clear();
        calibrating_baseline_ = true;
        messages_received_ = 0;

        // Reset Kalman filters
        for (auto &kf : kalman_filters_)
        {
            kf.reset();
        }

        // Publish calibration status
        auto status_msg = std_msgs::msg::Bool();
        status_msg.data = true;
        calibration_status_pub_->publish(status_msg);

        response->success = true;
        response->message = "Baseline recalibration started.";
        response->wait_time_ms = 26 * baseline_sample_size_;
    }

    void jointStateCallback(const sensor_msgs::msg::JointState::SharedPtr msg)
    {
        // Decimate input frequency: only process every Nth message
        callback_counter_++;
        if (callback_counter_ % process_decimation_ != 0)
        {
            return;
        }

        auto current_time = std::chrono::high_resolution_clock::now();

        // Track timing
        if (last_callback_time_.time_since_epoch().count() > 0)
        {
            auto dt = std::chrono::duration<double>(current_time - last_callback_time_).count();
            if (dt > 0)
            {
                callback_dt_history_.push_back(dt);
                if (callback_dt_history_.size() > 100)
                {
                    callback_dt_history_.pop_front();
                }
            }
        }
        last_callback_time_ = current_time;
        messages_received_++;

        // Check message validity
        if (msg->effort.size() < 6 || msg->position.size() < 6)
        {
            return;
        }

        // Reorder joints (matching Python: [5, 0, 1, 2, 3, 4])
        std::vector<int> reorder = {5, 0, 1, 2, 3, 4};
        std::vector<double> torques(6), positions(6);
        for (int i = 0; i < 6; ++i)
        {
            torques[i] = msg->effort[reorder[i]];
            positions[i] = msg->position[reorder[i]];
        }

        current_joint_angles_ = positions;

        // Filter torques
        std::vector<double> filtered_torques;
        for (int i = 0; i < 6; ++i)
        {
            filtered_torques.push_back(kalman_filters_[i].update(torques[i]));
        }

        if (calibrating_baseline_)
        {
            handleCalibration(filtered_torques);
        }
        else
        {
            estimateMassPhysics(filtered_torques);
        }
    }

    void handleCalibration(const std::vector<double> &filtered_torques)
    {
        std::vector<double> active_torques;
        for (auto idx : active_joints_)
        {
            active_torques.push_back(filtered_torques[idx]);
        }

        baseline_samples_.push_back(active_torques);

        if (static_cast<int>(baseline_samples_.size()) >= baseline_sample_size_)
        {
            // Calculate median baseline for each active joint
            baseline_torques_ = calculateMedian(baseline_samples_);
            calibrating_baseline_ = false;

            auto status_msg = std_msgs::msg::Bool();
            status_msg.data = false;
            calibration_status_pub_->publish(status_msg);

            RCLCPP_INFO(this->get_logger(), "Baseline calibration complete");

            std::string baseline_str = "Baseline torques: [";
            for (size_t i = 0; i < baseline_torques_.size(); ++i)
            {
                baseline_str += std::to_string(baseline_torques_[i]);
                if (i < baseline_torques_.size() - 1)
                    baseline_str += ", ";
            }
            baseline_str += "]";
            RCLCPP_INFO(this->get_logger(), "%s", baseline_str.c_str());

            // Report callback rate
            if (!callback_dt_history_.empty())
            {
                double avg_dt = std::accumulate(callback_dt_history_.begin(),
                                                callback_dt_history_.end(), 0.0) /
                                callback_dt_history_.size();
                RCLCPP_INFO(this->get_logger(),
                            "Actual callback rate: %.1f Hz", 1.0 / avg_dt);
            }
        }
    }

    void estimateMassPhysics(const std::vector<double> &filtered_torques)
    {
        if (baseline_torques_.empty() || current_joint_angles_.empty())
        {
            return;
        }

        Eigen::VectorXd torque_deltas(active_joints_.size());
        for (size_t i = 0; i < active_joints_.size(); ++i)
        {
            torque_deltas(i) = filtered_torques[active_joints_[i]] - baseline_torques_[i];
        }

        Eigen::Vector3d all_moment_arms = kinematics_.computeMomentArms(current_joint_angles_);

        Eigen::VectorXd active_moment_arms(active_joints_.size());
        for (size_t i = 0; i < active_joints_.size(); ++i)
        {
            active_moment_arms(i) = all_moment_arms(active_joints_[i] - 1);
        }

        double raw_mass = std::abs(kinematics_.estimateMass(torque_deltas, active_moment_arms));
        double raw_grams = raw_mass * 1000.0;

        int output_mass = 0;

        RCLCPP_DEBUG_THROTTLE(this->get_logger(), *this->get_clock(), 1000,
                              "Raw mass: %.4f grams", raw_grams);

        if (use_snapping_)
        {
            // Exponential calibration mode + snap to weight set (0, 50, 100, 200, 500)
            if (raw_mass < min_threshold_)
            {
                calibration_factor_ = 0.0;
            }
            else if (raw_mass < mass_threshold_)
            {
                calibration_factor_ = exp_amplitude_light_ * std::exp(-decay_light_ * raw_mass);
            }
            else
            {
                calibration_factor_ = exp_amplitude_heavy_ * std::exp(-decay_heavy_ * raw_mass);
            }

            estimated_mass_grams_ = raw_mass * calibration_factor_ * 1000.0;
            estimated_mass_grams_ = std::round(estimated_mass_grams_ / 5.0) * 5.0;

            output_mass = snapToWeightSet(estimated_mass_grams_);
        }
        else
        {
            // Polynomial calibration mode + round to nearest 5g (no snapping)
            double calibrated_grams = 0.0;
            double min_threshold_grams_ = min_threshold_ * 1000.0;

            if (raw_grams >= min_threshold_grams_)
            {
                if (raw_grams > 80.0)
                {
                    // Linear extrapolation for high values (during transients)
                    // Use gradient from 500g region
                    calibrated_grams = 500.0 + (raw_grams - 68.4) * 4.0;
                }
                else
                {
                    calibrated_grams = poly_coeff_b_ * raw_grams * raw_grams + poly_coeff_c_ * raw_grams + poly_coeff_d_;
                }
            }

            estimated_mass_grams_ = std::max(0.0, calibrated_grams);
            estimated_mass_grams_ = std::round(estimated_mass_grams_ / 5.0) * 5.0;

            output_mass = static_cast<int>(estimated_mass_grams_);
        }

        auto mass_msg = std_msgs::msg::Int32();
        mass_msg.data = output_mass;
        mass_pub_->publish(mass_msg);
    }

    int snapToWeightSet(double estimated_mass)
    {
        // Range-based snapping
        if (estimated_mass <= 17.5)
        {
            return 0;
        }
        else if (estimated_mass <= 45)
        {
            return 50;
        }
        else if (estimated_mass <= 185)
        {
            return 100;
        }
        else if (estimated_mass <= 375)
        {
            return 200;
        }
        else
        {
            return 500;
        }
    }

    std::vector<double> calculateMedian(const std::vector<std::vector<double>> &samples)
    {
        if (samples.empty())
            return {};

        size_t num_values = samples[0].size();
        std::vector<double> medians(num_values);

        for (size_t j = 0; j < num_values; ++j)
        {
            std::vector<double> column;
            for (const auto &sample : samples)
            {
                if (j < sample.size())
                {
                    column.push_back(sample[j]);
                }
            }

            if (!column.empty())
            {
                std::nth_element(column.begin(),
                                 column.begin() + column.size() / 2,
                                 column.end());
                medians[j] = column[column.size() / 2];
            }
        }

        return medians;
    }

    void reportDiagnostics()
    {
        if (!callback_dt_history_.empty() && messages_received_ > 0)
        {
            double avg_dt = std::accumulate(callback_dt_history_.begin(),
                                            callback_dt_history_.end(), 0.0) /
                            callback_dt_history_.size();
            RCLCPP_DEBUG(this->get_logger(),
                         "Diagnostics: %zu messages, callback rate: %.1f Hz",
                         messages_received_, 1.0 / avg_dt);
        }
    }

    // ROS interfaces
    rclcpp::Subscription<sensor_msgs::msg::JointState>::SharedPtr joint_state_sub_;
    rclcpp::Publisher<std_msgs::msg::Int32>::SharedPtr mass_pub_;
    rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr calibration_status_pub_;
    rclcpp::Service<sort_interfaces::srv::CalibrateBaseline>::SharedPtr calibration_service_;
    rclcpp::TimerBase::SharedPtr diagnostic_timer_;
    rclcpp::TimerBase::SharedPtr param_timer_;

    // Kalman filters and kinematics
    std::vector<KalmanFilter> kalman_filters_;
    UR5eKinematics kinematics_;

    // Timing
    std::deque<double> callback_dt_history_;
    std::chrono::high_resolution_clock::time_point last_callback_time_;

    // State
    bool calibrating_baseline_;
    std::vector<std::vector<double>> baseline_samples_;
    std::vector<double> baseline_torques_;
    std::vector<double> current_joint_angles_;
    std::vector<int64_t> active_joints_;

    // Calibration parameters
    bool use_snapping_; // When false, uses polynomial calibration (equivalent to deprecated use_polynomial_calibration=true)
    double calibration_factor_;

    // Exponential calibration
    double exp_amplitude_light_;
    double exp_amplitude_heavy_;
    double decay_light_;
    double decay_heavy_;
    double mass_threshold_;
    double min_threshold_;

    // Polynomial calibration
    double poly_coeff_a_;
    double poly_coeff_b_;
    double poly_coeff_c_;
    double poly_coeff_d_;

    int baseline_sample_size_;

    double estimated_mass_grams_;
    size_t messages_received_;

    std::vector<int> weight_set_;

    // Input decimation
    size_t callback_counter_;
    int process_decimation_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<WeightDetector>());
    rclcpp::shutdown();
    return 0;
}
